osu.Game.Tests/Visual/Gameplay/TestSceneStoryboardWithOutro.cs

```diff
@@ -0,0 +1,113 @@
+            sprite.TimelineGroup.Alpha.Add(Easing.None, 0, storyboard_duration, 1, 0);
+            storyboard.GetLayer("Background").Add(sprite);
+        }
+
+        [SetUpSteps]
+        public override void SetUpSteps()
+        {
+            base.SetUpSteps();
+            AddStep("ignore user settings", () =>
+            {
+                Player.DimmableStoryboard.IgnoreUserSettings.Value = true;
+            });
+        }
+
+        [Test]
+        public void TestStoryboardSkipOutro()
+        {
+            AddUntilStep("storyboard loaded", () => Player.Beatmap.Value.StoryboardLoaded);
+            AddUntilStep("completion set by processor", () => Player.ScoreProcessor.HasCompleted.Value);
+            AddStep("skip outro", () => InputManager.Key(osuTK.Input.Key.Space));
+            AddAssert("score shown", () => Player.IsScoreShown);
+        }
+
+        [Test]
+        public void TestStoryboardNoSkipOutro()
+        {
+            AddUntilStep("storyboard loaded", () => Player.Beatmap.Value.StoryboardLoaded);
+            AddUntilStep("storyboard ends", () => Player.HUDOverlay.Progress.ReferenceClock.CurrentTime >= storyboard_duration);
+            AddUntilStep("wait for score shown", () => Player.IsScoreShown);
+        }
+
+        [Test]
+        public void TestStoryboardExitToSkipOutro()
+        {
+            AddUntilStep("storyboard loaded", () => Player.Beatmap.Value.StoryboardLoaded);
+            AddUntilStep("completion set by processor", () => Player.ScoreProcessor.HasCompleted.Value);
+            AddStep("exit via pause", () => Player.ExitViaPause());
+            AddAssert("score shown", () => Player.IsScoreShown);
+        }
+
+        protected override bool AllowFail => false;
+
+        protected override Ruleset CreatePlayerRuleset() => new OsuRuleset();
+
+        protected override TestPlayer CreatePlayer(Ruleset ruleset) => new OutroPlayer();
+
+        protected override IBeatmap CreateBeatmap(RulesetInfo ruleset)
+        {
+            var beatmap = new Beatmap();
+            beatmap.HitObjects.Add(new HitCircle());
+            return beatmap;
+        }
+
+        protected override WorkingBeatmap CreateWorkingBeatmap(IBeatmap beatmap, Storyboard storyboard = null) =>
+            new ClockBackedTestWorkingBeatmap(beatmap, storyboard ?? this.storyboard, Clock, Audio);
+
+        protected class OutroPlayer : TestPlayer
+        {
+            public void ExitViaPause() => PerformExit(true);
+
+            public bool IsScoreShown => !this.IsCurrentScreen() && this.GetChildScreen() is ResultsScreen;
+
+            public new DimmableStoryboard DimmableStoryboard => base.DimmableStoryboard;
+
+            public OutroPlayer()
+                : base(false)
+            {
+            }
+
+            protected override Task ImportScore(Score score)
+            {
+                // avoid database errors from trying to store the score
+                return Task.CompletedTask;
+            }
```

Comment:
```
[database] 2021-04-17 16:34:08 [verbose]: [?????] Beginning import...
[database] 2021-04-17 16:34:10 [error]: Failed executing DbCommand (8ms) [Parameters=[@p10='?', @p11='?', @p12='?', @p13='?', @p14='?', @p15='?', @p16='?', @p17='?', @p18='?', @p19='?', @p20='?', @p21='?', @p22='?', @p23='?', @p24='?', @p25='?', @p26='?', @p27='?', @p28='?', @p29='?', @p30='?', @p31='?', @p32='?', @p33='?', @p34='?' (Size = 6), @p35='?'], CommandType='Text', CommandTimeout='10']
[database] 2021-04-17 16:34:10 [error]: INSERT INTO "BeatmapInfo" ("AudioLeadIn", "BPM", "BaseDifficultyID", "BeatDivisor", "BeatmapSetInfoID", "Countdown", "DistanceSpacing", "EpilepsyWarning", "GridSize", "Hash", "Hidden", "Length", "LetterboxInBreaks", "MD5Hash", "MetadataID", "OnlineBeatmapID", "Path", "RulesetID", "SpecialStyle", "StackLeniency", "StarDifficulty", "Status", "StoredBookmarks", "TimelineZoom", "Version", "WidescreenStoryboard")
[database] 2021-04-17 16:34:10 [error]: VALUES (@p10, @p11, @p12, @p13, @p14, @p15, @p16, @p17, @p18, @p19, @p20, @p21, @p22, @p23, @p24, @p25, @p26, @p27, @p28, @p29, @p30, @p31, @p32, @p33, @p34, @p35);
[database] 2021-04-17 16:34:10 [error]: SELECT "ID"
[database] 2021-04-17 16:34:10 [error]: FROM "BeatmapInfo"
[database] 2021-04-17 16:34:10 [error]: WHERE changes() = 1 AND "ID" = last_insert_rowid();
[database] 2021-04-17 16:34:10 [error]: An exception occurred in the database while saving changes for context type 'osu.Game.Database.OsuDbContext'.
[database] 2021-04-17 16:34:10 [error]: Microsoft.EntityFrameworkCore.DbUpdateException: An error occurred while updating the entries. See the inner exception for details.
[database] 2021-04-17 16:34:10 [error]: ---> Microsoft.Data.Sqlite.SqliteException (0x80004005): SQLite Error 19: 'FOREIGN KEY constraint failed'.
[database] 2021-04-17 16:34:10 [error]: at Microsoft.Data.Sqlite.SqliteException.ThrowExceptionForRC(Int32 rc, sqlite3 db)
[database] 2021-04-17 16:34:10 [error]: at Microsoft.Data.Sqlite.SqliteCommand.ExecuteReader(CommandBehavior behavior)
[database] 2021-04-17 16:34:10 [error]: at Microsoft.Data.Sqlite.SqliteCommand.ExecuteDbDataReader(CommandBehavior behavior)
[database] 2021-04-17 16:34:10 [error]: at System.Data.Common.DbCommand.ExecuteReader()
[database] 2021-04-17 16:34:10 [error]: at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.Execute(IRelationalConnection connection, DbCommandMethod executeMethod, IReadOnlyDictionary`2 parameterValues)
[database] 2021-04-17 16:34:10 [error]: at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.ExecuteReader(IRelationalConnection connection, IReadOnlyDictionary`2 parameterValues)
[database] 2021-04-17 16:34:10 [error]: at Microsoft.EntityFrameworkCore.Update.ReaderModificationCommandBatch.Execute(IRelationalConnection connection)
[database] 2021-04-17 16:34:10 [error]: --- End of inner exception stack trace ---
```

I believe it's due to importing a score info of a beatmap which doesn't exist in the database, that's the main reason I've suggested moving this at `TestPlayer`, as mostly all of them contain programmatically written beatmaps.

And in the end the scores will be imported to your actual database when running this in visual tests, and I don't think that's really a wanted behaviour. If anything, it could be a toggleable property if some tests want to test the score import process.

Reply:
> I believe it's due to importing a score info of a beatmap which doesn't exist in the database

that doesn't mean that this will always be the case. some test cases want to import actual maps and import actual scores. i'd rather this behaviour of crippling imports be opt-out and not opt-in. more explicit, less of silent "things are disabled in tests by default" hackery.

> And in the end the scores will be imported to your actual database when running this in visual tests

not necessarily? would recommend checking out the `bin/Debug/net5.0/` directory in a test project. in particular you should be seeing separate `client.db` files.

Reply:
I got FOREIGN KEY constraint failed without the suppression. I think it's because the game was trying to import the fake beatmap I created, while other tests use the `TestBeatmap` to completion or don't play out a fake beatmap to the score screen (`TestSceneCompletionCancellation`)

Here's part of the stack trace:

```
[database] 2021-04-17 16:42:10 [error]: Failed executing DbCommand (0ms) [Parameters=[@p10='?', @p11='?', @p12='?', @p13='?', @p14='?', @p15='?', @p16='?', @p17='?', @p18='?', @p19=
'?', @p20='?', @p21='?', @p22='?', @p23='?', @p24='?', @p25='?', @p26='?', @p27='?', @p28='?', @p29='?', @p30='?', @p31='?', @p32='?', @p33='?', @p34='?' (Size = 6), @p35='?'], Comm
andType='Text', CommandTimeout='10']
[database] 2021-04-17 16:42:10 [error]: INSERT INTO "BeatmapInfo" ("AudioLeadIn", "BPM", "BaseDifficultyID", "BeatDivisor", "BeatmapSetInfoID", "Countdown", "DistanceSpacing", "Epil
epsyWarning", "GridSize", "Hash", "Hidden", "Length", "LetterboxInBreaks", "MD5Hash", "MetadataID", "OnlineBeatmapID", "Path", "RulesetID", "SpecialStyle", "StackLeniency", "StarDif
ficulty", "Status", "StoredBookmarks", "TimelineZoom", "Version", "WidescreenStoryboard")
[database] 2021-04-17 16:42:10 [error]: VALUES (@p10, @p11, @p12, @p13, @p14, @p15, @p16, @p17, @p18, @p19, @p20, @p21, @p22, @p23, @p24, @p25, @p26, @p27, @p28, @p29, @p30, @p31, @
p32, @p33, @p34, @p35);
[database] 2021-04-17 16:42:10 [error]: SELECT "ID"
[database] 2021-04-17 16:42:10 [error]: FROM "BeatmapInfo"
[database] 2021-04-17 16:42:10 [error]: WHERE changes() = 1 AND "ID" = last_insert_rowid();
[database] 2021-04-17 16:42:10 [error]: An exception occurred in the database while saving changes for context type 'osu.Game.Database.OsuDbContext'.
[database] 2021-04-17 16:42:10 [error]: Microsoft.EntityFrameworkCore.DbUpdateException: An error occurred while updating the entries. See the inner exception for details.
[database] 2021-04-17 16:42:10 [error]: ---> Microsoft.Data.Sqlite.SqliteException (0x80004005): SQLite Error 19: 'FOREIGN KEY constraint failed'.
[database] 2021-04-17 16:42:10 [error]: at Microsoft.Data.Sqlite.SqliteException.ThrowExceptionForRC(Int32 rc, sqlite3 db)
[database] 2021-04-17 16:42:10 [error]: at Microsoft.Data.Sqlite.SqliteCommand.ExecuteReader(CommandBehavior behavior)
[database] 2021-04-17 16:42:10 [error]: at Microsoft.Data.Sqlite.SqliteCommand.ExecuteDbDataReader(CommandBehavior behavior)
[database] 2021-04-17 16:42:10 [error]: at System.Data.Common.DbCommand.ExecuteReader()
[database] 2021-04-17 16:42:10 [error]: at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.Execute(IRelationalConnection connection, DbCommandMethod executeMethod,
IReadOnlyDictionary`2 parameterValues)
[database] 2021-04-17 16:42:10 [error]: at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.ExecuteReader(IRelationalConnection connection, IReadOnlyDictionary`2 par
ameterValues)
[database] 2021-04-17 16:42:10 [error]: at Microsoft.EntityFrameworkCore.Update.ReaderModificationCommandBatch.Execute(IRelationalConnection connection)
[database] 2021-04-17 16:42:10 [error]: --- End of inner exception stack trace ---
```

Reply:
> > I believe it's due to importing a score info of a beatmap which doesn't exist in the database
> 
> that doesn't mean that this will always be the case. some test cases want to import actual maps and import actual scores. i'd rather this behaviour of crippling imports be opt-in and not opt-out. more explicit, less of silent "things are disabled in tests by default" hackery.

I guess you have a point on this one.

> > And in the end the scores will be imported to your actual database when running this in visual tests
> 
> not necessarily? would recommend checking out the `bin/Debug/net5.0/` directory in a test project. in particular you should be seeing separate `client.db` files for particular test scene executions.

I believe the separate `client.db` files are for NUnit test executions, not visual tests browser:
![CleanShot 2021-04-17 at 20 07 54](https://user-images.githubusercontent.com/22781491/115120935-c56d3700-9fb8-11eb-9607-05c184df31fc.png)